src/unix/lwt_main.ml

```diff
@@ -20,28 +20,84 @@ let yielded = Lwt_sequence.create ()
 
 let yield () = (Lwt.add_task_r [@ocaml.warning "-3"]) yielded
 
-let rec run t =
-  (* Wakeup paused threads now. *)
-  Lwt.wakeup_paused ();
-  match Lwt.poll t with
-  | Some x ->
-    x
-  | None ->
-    (* Call enter hooks. *)
-    Lwt_sequence.iter_l (fun f -> f ()) enter_iter_hooks;
-    (* Do the main loop call. *)
-    Lwt_engine.iter (Lwt.paused_count () = 0 && Lwt_sequence.is_empty yielded);
-    (* Wakeup paused threads again. *)
+let run_already_called = ref `No
+let run_already_called_mutex = Mutex.create ()
+
+let run p =
+  (* Fail in case a call to Lwt_main.run is nested under another invocation of
+     Lwt_main.run. *)
+  Mutex.lock run_already_called_mutex;
+
+  let error_message_if_call_is_nested =
+    match !run_already_called with
+    | `From backtrace_string ->
+      Some (Printf.sprintf "%s\n%s\n%s"
+        "Nested calls to Lwt_main.run are not allowed"
+        "Lwt_main.run already called from:"
+        backtrace_string)
+    | `From_somewhere ->
+      Some ("Nested calls to Lwt_main.run are not allowed")
+    | `No ->
+      let called_from =
+        if Printexc.backtrace_status () then
+          let backtrace =
+            try raise Exit
+            with Exit -> Printexc.get_backtrace ()
+          in
```

Comment:
I decided to try recording stack traces of where `Lwt_main.run` was called, rather than add an optional argument.